state/factory/workingset.go

```diff
@@ -404,6 +404,40 @@ func (ws *workingSet) process(ctx context.Context, actions []action.SealedEnvelo
 	return ws.finalize()
 }
 
+func (ws *workingSet) validateSystemActionLayout(ctx context.Context, actions []action.SealedEnvelope) error {
+	reg := protocol.MustGetRegistry(ctx)
+	postSystemActions := []action.Envelope{}
+	for _, p := range reg.All() {
+		if psc, ok := p.(protocol.PostSystemActionsCreator); ok {
+			elps, err := psc.CreatePostSystemActions(ctx, ws)
+			if err != nil {
+				return err
+			}
+			postSystemActions = append(postSystemActions, elps...)
+		}
+	}
+	receivedSystemActions := []action.Envelope{}
+	startIdx := len(actions)
+	for i := range actions {
+		if action.IsSystemAction(actions[i]) {
+			if startIdx == len(actions) {
+				startIdx = i
+			}
+			receivedSystemActions = append(receivedSystemActions, actions[i].Envelope)
+		}
+	}
```

Comment:
system actions should be at the end of the action list, and they should be continuous.